optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java

```diff
@@ -0,0 +1,371 @@
+                .penalize("Room penalty exclusive", HardSoftScore.ONE_HARD,
+                          (periodPenalty, leftExam, rightExam) -> leftExam.getTopic().getStudentSize() + rightExam.getTopic().getStudentSize());
+    }
+
+    //    // Two exams in a row which share students
+//    rule "twoExamsInARow"
+//    when
+//    $topicConflict : TopicConflict($leftTopic : leftTopic, $rightTopic : rightTopic)
+//    $leftExam : Exam(topic == $leftTopic, $leftDayIndex : dayIndex, $leftPeriodIndex : periodIndex, period != null)
+//    $rightExam : Exam(topic == $rightTopic, dayIndex == $leftDayIndex,
+//                      Math.abs($leftPeriodIndex - periodIndex) == 1)
+//    then
+//        scoreHolder.penalize(kcontext, $topicConflict.getStudentSize());
+//    end
+    private Constraint twoExamsInARow(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(TopicConflict.class)
+                .join(Exam.class,
+                      equal(TopicConflict::getLeftTopic, Exam::getTopic))
+                .filter((topicConflict, leftExam) -> leftExam.getPeriod() != null)
+                .join(Exam.class,
+                      equal((topicConflict, leftExam) -> topicConflict.getRightTopic(), Exam::getTopic),
+                      equal((topicConflict, leftExam) -> leftExam.getDayIndex(), Exam::getDayIndex))
+                .filter((topicConflict, leftExam, rightExam) -> Math.abs(leftExam.getPeriodIndex() - rightExam.getPeriodIndex()) == 1)
+                .penalize("Two exams in a row", HardSoftScore.ONE_SOFT, (topicConflict, leftExam, rightExam) -> topicConflict.getStudentSize());
+    }
+
+    //    // Two exams in a day which share students
+//    rule "twoExamsInADay"
+//    when
+//    $topicConflict : TopicConflict($leftTopic : leftTopic, $rightTopic : rightTopic)
+//    $leftExam : Exam(topic == $leftTopic, $leftDayIndex : dayIndex, $leftPeriodIndex : periodIndex, period != null)
+//    $rightExam : Exam(topic == $rightTopic, dayIndex == $leftDayIndex,
+//                      Math.abs($leftPeriodIndex - periodIndex) > 1)
+//    then
+//        scoreHolder.penalize(kcontext, $topicConflict.getStudentSize());
+//    end
+    private Constraint twoExamsInADay(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(TopicConflict.class)
+                .join(Exam.class,
+                      equal(TopicConflict::getLeftTopic, Exam::getTopic))
+                .filter((topicConflict, leftExam) -> leftExam.getPeriod() != null)
+                .join(Exam.class,
+                      equal((topicConflict, leftExam) -> topicConflict.getRightTopic(), Exam::getTopic),
+                      equal((topicConflict, leftExam) -> leftExam.getDayIndex(), Exam::getDayIndex))
+                .filter((topicConflict, leftExam, rightExam) -> Math.abs(leftExam.getPeriodIndex() - rightExam.getPeriodIndex()) > 1)
```

Comment:
Although this isn't functionally equivalent to the DRL, you could get away with replacing those joins with two `ifExists()`. Good for performance, but may not pass `FULL_ASSERT`.